lookout/style/format/feature_extractor.py

```diff
@@ -796,15 +795,130 @@ def _compute_labels_mappings(self, vnodes: Iterable[VirtualNode]) -> None:
+            if annotation.has_node:
+                closest_left_node_id = id(annotation.node)
+                file.add(TokenParentAnnotation(*annotation.span,
+                                               uast_annotation.parents[closest_left_node_id]))
             else:
-                parent = (self._find_parent(j, file_vnodes, file_parents, closest_left_node_id)
-                          or uast)
-                vnode_parents[id(vn)] = parent
+                parent = (self._find_parent(annotation.stop, file, closest_left_node_id)
+                          or uast_annotation.uast)
+                file.add(TokenParentAnnotation(*annotation.span, parent))
+
+
+def _to_position(raw_lines_data, _lines_start_offset, offset):
+    line_num = numpy.argmax(_lines_start_offset > offset) - 1
+    col = offset - _lines_start_offset[line_num]
+    line = raw_lines_data[line_num]
+    if len(line) == col:
+        if line.splitlines()[0] != line:
+            # ends with newline
+            line_num += 1
+            col = 0
+    return Position(offset, line_num + 1, col + 1)
+
+
+def raw_file_to_vnodes_and_parents(file: AnnotationManager) -> Tuple[List["VirtualNode"],
+                                                                     Dict[int, bblfsh.Node]]:
+    """
+    Convert `AnnotationManager` instance to the deprecated output format of \
+    `FeatureExtractor._parse_file()`.
+
+    The function is created for backwards capability and should be removed after refactoring is \
+    finished.
+
+    :param file: file annotated with `UASTAnnotation`, `PathAnnotation` and `RawTokenAnnotation`. \
+                 It is expected to be the output of  `FeatureExtractor._parse_file()`.
+    :return: The old `FeatureExtractor._parse_file()` output format, that is \
+             Tuple with `VirtualNode`-s and `bbfsh.Node` id to parent mapping.
+    """
+    vnodes = []
+    path = file.get(PathAnnotation).path
+    raw_lines_data = file.sequence.splitlines(keepends=True)
+    line_lens = [0] + [len(d) for d in raw_lines_data]
+    line_lens[-1] += 1
+    _line_start_offsets = numpy.array(line_lens).cumsum()
+    for annotation in file.iter_annotation(RawTokenAnnotation):
+        vnode = VirtualNode(
+            file[annotation.span],
+            _to_position(raw_lines_data, _line_start_offsets, annotation.start),
+            _to_position(raw_lines_data, _line_start_offsets, annotation.stop),
+            is_accumulated_indentation=False,
+            path=path,
+            node=annotation.node,
+            y=None,
+        )
+        vnodes.append(vnode)
+    return vnodes, file.get(UASTAnnotation).parents
+
+
+def file_to_vnodes_and_parents(file: AnnotationManager) -> Tuple[List["VirtualNode"],
+                                                                 Dict[int, bblfsh.Node]]:
+    """
+    Convert `AnnotationManager` instance to an old format, that is a sequence of vnodes and \
+    vnodes parents mapping.
```

Comment:
Following to the docstrings, I am really confused whether `raw_file`- and `file`- are the same or not. Either rename them or write more detailed docstrings which explain the difference.

Reply:
different functions for backwards capability of different `FeatureExtractor` methods. 
I add the main difference to a function name. 
Now we have 
`files_to_old_parse_vnodes_format`
`file_to_old_parse_file_format`
and internal 
`_file_to_vnodes_and_parents`

docstrings are also extended.